instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/EvRerouteTest.kt

```diff
@@ -0,0 +1,217 @@
+        }
+
+        // start a route
+        runOnMainSync {
+            mapboxNavigation.historyRecorder.startRecording()
+            mapboxNavigation.startTripSession()
+            mapboxNavigation.requestRoutes(
+                RouteOptions.builder()
+                    .applyDefaultNavigationOptions()
+                    .applyLanguageAndVoiceUnitOptions(activity)
+                    .baseUrl(mockWebServerRule.baseUrl)
+                    .coordinatesList(mockRoute.routeWaypoints)
+                    .unrecognizedProperties(
+                        mapOf(
+                            "engine" to "electric",
+                            "ev_max_charge" to "8000",
+                            "ev_connector_types" to "tesla",
+                            "energy_consumption_curve" to "0,300;20,160;80,140;120,180",
+                            "ev_charging_curve" to "40000,70000",
+                        )
+                    )
+                    .build(),
+                object : NavigationRouterCallback {
+                    override fun onRoutesReady(
+                        routes: List<NavigationRoute>,
+                        routerOrigin: RouterOrigin
+                    ) {
+                        mapboxNavigation.setNavigationRoutes(routes)
+                    }
+
+                    override fun onFailure(
+                        reasons: List<RouterFailure>,
+                        routeOptions: RouteOptions
+                    ) {
+                        logE("onFailure reasons=$reasons", "DEBUG")
+                    }
+
+                    override fun onCanceled(
+                        routeOptions: RouteOptions,
+                        routerOrigin: RouterOrigin
+                    ) {
+                        logE("onCanceled", "DEBUG")
+                    }
+                }
+            )
+        }
+
+        // wait for tracking to start
+        mapboxHistoryTestRule.stopRecordingOnCrash("no location tracking") {
+            Espresso.onIdle()
+        }
+        locationTrackingIdlingResource.unregister()
+
+        // push off route location and wait for the off route event
+        offRouteIdlingResource.register()
+        runOnMainSync {
+            mockLocationReplayerRule.loopUpdate(offRouteLocationUpdate, times = 5)
+        }
+
+        mapboxHistoryTestRule.stopRecordingOnCrash("no off route") {
+            Espresso.onIdle()
+        }
+        offRouteIdlingResource.unregister()
+
+        // wait for tracking to start again
+        locationTrackingIdlingResource.register()
+        mapboxHistoryTestRule.stopRecordingOnCrash("no tracking") {
+            Espresso.onIdle()
+        }
+        locationTrackingIdlingResource.unregister()
+
+        val countDownLatch = CountDownLatch(1)
+        runOnMainSync {
+            mapboxNavigation.historyRecorder.stopRecording {
+                logE("history path=$it", "DEBUG")
+                countDownLatch.countDown()
```

Comment:
It would be nice to introduce a suspend version of `stopRecording` because it's used a lot and every time we have to create a  `CountDownLatch`. But that's out of scope of this PR.